plasma_framework/python_tests/plasma_core/utils/eip712_struct_hash.py

```diff
@@ -1,63 +1,81 @@
+    if domain and verifying_contract:
         raise RuntimeError("verifyingContract supplied but ignored")
 
-    verifying_address = verifyingContract.address if verifyingContract else NULL_ADDRESS
-
-    inputs = [Input(blknum=i.blknum, txindex=i.txindex, oindex=i.oindex) for i in tx.inputs]
-    outputs = [Output(owner=o.output_guard, currency=o.token, amount=o.amount) for o in tx.outputs]
+    verifying_address = hex_to_binary(verifying_contract.address) if verifying_contract else NULL_ADDRESS
 
     domain = domain or make_domain(
         name='OMG Network',
         version='1',
         verifyingContract=verifying_address,
-        salt=bytes.fromhex('fad5c7f626d80f9256ef01929f3beb96e058b8b4b0e3fe52d84f054c0e2a7a83')
+        salt=hex_to_binary('fad5c7f626d80f9256ef01929f3beb96e058b8b4b0e3fe52d84f054c0e2a7a83')
     )
 
-    type = Transaction().encode_type() + Input.encode_type() + Output.encode_type()
-    values = [_hash_typed(t) for t in inputs] + [_hash_typed(t) for t in outputs] + [tx.metadata or NULL_HASH]
-
-    return keccak(b'\x19\x01' + _hash_typed(domain) + _hash_struct(type, b''.join(values)))
-
-
-def _hash_struct(type, value):
-    return keccak(keccak(text=type) + value)
+    return keccak(b'\x19\x01' + domain.hash_struct() + struct_tx_from_tx(tx).hash_struct())
 
 
-def _hash_typed(obj):
-    return _hash_struct(obj.encode_type(), obj.encode_value())
-
-
-# NOTE: eip712_structs computes hashes incorrectly however we could fix implementation based on it's building blocks
 class Input(EIP712Struct):
     blknum = Uint(256)
     txindex = Uint(256)
     oindex = Uint(256)
 
 
 class Output(EIP712Struct):
-    output_guard = Bytes(20)
+    outputType = Uint(256)
+    outputGuard = Bytes(20)
     currency = Address()
     amount = Uint(256)
 
 
 class Transaction(EIP712Struct):
-    input0 = Input()
-    input1 = Input()
-    input2 = Input()
-    input3 = Input()
-    output0 = Output()
-    output1 = Output()
-    output2 = Output()
-    output3 = Output()
+    txType = Uint(256)
+    input0 = Input
+    input1 = Input
+    input2 = Input
+    input3 = Input
+    output0 = Output
+    output1 = Output
+    output2 = Output
+    output3 = Output
     metadata = Bytes(32)
+
+
+def struct_tx_from_tx(tx):
+    inputs = _map_inputs(tx.inputs)
+    outputs = _map_outputs(tx.outputs)
+
+    return Transaction(
+        txType=tx.tx_type,
+        input0=inputs[0],
+        input1=inputs[1],
+        input2=inputs[2],
+        input3=inputs[3],
+        output0=outputs[0],
+        output1=outputs[1],
+        output2=outputs[2],
+        output3=outputs[3],
+        metadata=tx.metadata,
+    )
+
+
+def _map_inputs(inputs):
+    empty_input = Input()
+
+    eip712_inputs = [Input(blknum=i.blknum, txindex=i.txindex, oindex=i.oindex) for i in inputs]
+    return eip712_inputs + [empty_input] * (4 - len(inputs))  # pad with empty inputs
+
+
+def _map_outputs(outputs):
+    empty_output = Output()
+
+    eip712_outputs = [
```

Comment:
minor: might consider to just write in `for` style as it is using multiple lines anyway. We can even just pad the empty output within the loop too.